### DISCUSSION TITLE
Does an NPR_Specular node slow my rendering if I give an empty gradient input?

### DISCUSSION THREAD

**Discussion by HSDSZ** [bnpr/Malt#383] [Q&A]
![image](https://user-images.githubusercontent.com/36077492/185758183-febf5ce4-df91-45d1-9f37-c9a957db8ca1.png)

Here I have two node added and output to the color. 
If I set the gradient of NPR_Specular empty, the result is the same as the one with only NPR_Diffuse node.
Will the NPR_Sepcular slow my rendering if it is empty?

**Comment by HSDSZ** [comment]
Another one
I connect an image texture to the prepass ID to render custom lines
![image](https://user-images.githubusercontent.com/36077492/185792366-0f23cecc-52b7-41b1-a34d-46fb6cee6abe.png)

If I don't set any image, will the rendering be slowed?

**Comment by pragma37** [comment]
I can't give you a "real" answer.

Any node that gets connected to the final output will generate extra code that is supposed to be executed, regardless of its parameters.
I think your driver could theoretically be clever enough to generate shader specializations at runtime and optimize out "death" code paths.
Another question would be if those 2 additions, even when executed, take a meaningful amount of the total render time.

So the only way to know the answer is to connect and disconnect those nodes in a real scene and see if they have a significant performance impact.  ¯\_(ツ)_/¯

**Reply by pragma37** [reply]
BTW, in that second screenshot you are overriding the first ID channel, so "self-shadow" detection is not going to work correctly.
